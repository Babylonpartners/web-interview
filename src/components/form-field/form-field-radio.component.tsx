import React, { Component, ReactNode } from 'react'
import dayjs from 'dayjs'

import { Icon } from 'components/icon/icon.component'

import './form-field.component.scss'

type FieldProps = {
  label: string
  icon: string
  options?: any[]
  noOptionsText?: string
  placeholder?: string
  onChange: (type: string, data: string) => void
}

type FieldState = {
  value: any
  options?: any[]
}

export class FormFieldRadio extends Component<FieldProps, FieldState> {
  label: FieldProps['label']
  icon: FieldProps['icon']
  options?: FieldProps['options']
  placeholder?: FieldProps['placeholder']
  onChange: FieldProps['onChange']

  constructor(props: FieldProps) {
    super(props)

    this.state = {
      value: undefined,
      options: [],
    }

    this.label = props.label
    this.icon = props.icon
    this.onChange = props.onChange

    this.options = props.options || undefined
  }

  componentDidUpdate(prevProps: FieldProps, prevState: FieldState) {
    if (this.props.options !== prevState.options) {
      this.setState({ options: this.props.options })
    }
  }

  selectOption(option) {
    this.setState({ value: option })
    this.props.onChange(this.label, option)
  }

  renderOptions(
    options: FieldProps['options'],
    noOptionsText?: string
  ): ReactNode[] {
    if (options) {
      return options.map((option, key) => {
        return (
          <span
            key={key}
            className={`form-field__option pill${
              this.state.value === option ? ' pill--active' : ''
            }`}
            onClick={() => {
              this.selectOption(option)
            }}
          >
            {option.type === 'date'
              ? dayjs(option.value).format('dddd HH:mm')
              : option.value}
          </span>
        )
      })
    }

    if (noOptionsText) {
      return [
        <span key="0" className="form-field__option pill pill--disabled">
          {noOptionsText}
        </span>,
      ]
    }
  }

  render() {
    return (
      <div className="form-field form-field--radio">
        <div className="form-field__label">
          <Icon alias={this.icon}></Icon>
          <span>{this.label}</span>
        </div>

        <div className="form-field__input">
          {this.renderOptions(
            this.state.options,
            this.props.noOptionsText || undefined
          )}
        </div>
      </div>
    )
  }
}
