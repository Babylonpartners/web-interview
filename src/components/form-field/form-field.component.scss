@import 'assets/styles/_constants.scss';

.form-field {
  &__label {
    font-size: 11rem;
    font-weight: 600;
    color: $label-color;
  }

  textarea {
    padding: 7rem 8rem;
    border-radius: 5rem;
    border: 1px solid $outline-dark;
    min-width: 100%;
    max-width: 100%;
    min-height: 100px;
    max-height: 200px;
  }

  &__input {
    font-size: 11rem;
    margin-top: 16rem;

    &:focus {
      border-color: $focus-border;
      outline: none;
    }
  }

  &--radio {
    .form-field__input {
      margin: #{16rem - $gutter-size} -#{$gutter-size} -#{$gutter-size};
    }

    .form-field__option {
      margin: $gutter-size;
    }
  }
}
