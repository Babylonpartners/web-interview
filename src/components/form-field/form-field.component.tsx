import React, { Component } from 'react'
import { Icon } from 'components/icon/icon.component'

import './form-field.component.scss'

type FieldProps = {
  type: string
  label: string
  icon: string
  placeholder?: string
  onChange: (type: string, data: string) => void
}

type FieldState = {
  value: string
}

export class FormField extends Component<FieldProps, FieldState> {
  type: FieldProps['type']
  label: FieldProps['label']
  icon: FieldProps['icon']
  placeholder?: FieldProps['placeholder']
  onChange: FieldProps['onChange']

  constructor(props: FieldProps) {
    super(props)

    this.state = {
      value: undefined,
    }

    this.type = props.type
    this.label = props.label
    this.icon = props.icon
    this.onChange = props.onChange

    this.placeholder = props.placeholder || undefined
  }

  handleChange = event => {
    this.setState({ value: event.target.value })
    this.props.onChange(this.label, event.target.value)
  }

  render() {
    return (
      <div className={`form-field form-field--${this.type}`}>
        <div className="form-field__label">
          <Icon alias={this.icon}></Icon>
          <span>{this.label}</span>
        </div>

        {this.type === 'textarea' && (
          <textarea
            className="form-field__input"
            onChange={this.handleChange}
            placeholder={this.placeholder}
          ></textarea>
        )}

        {this.type === 'input' && (
          <input className="form-field__input" onChange={this.handleChange} />
        )}
      </div>
    )
  }
}
