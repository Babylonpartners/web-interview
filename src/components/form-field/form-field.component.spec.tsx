import React from 'react'
import renderer from 'react-test-renderer'
import ReactSixteenAdapter from 'enzyme-adapter-react-16'
import * as enzyme from 'enzyme'

enzyme.configure({ adapter: new ReactSixteenAdapter() })

import { FormField } from './form-field.component'

describe('FormField', () => {
  const onChangeHandler = jest.fn()
  const element = (
    <FormField
      type="input"
      label="text"
      icon="atom"
      onChange={onChangeHandler}
    />
  )

  let component

  beforeEach(() => {
    const wrapper = enzyme.shallow(element)

    component = wrapper.instance()
  })

  it('should match snapshot', () => {
    const snapshotComponent = renderer.create(element)
    expect(snapshotComponent.toJSON()).toMatchSnapshot()
  })

  it('updates the state and emits the value to the parent when an option is selected', () => {
    expect(component.state.value).toBeUndefined()

    component.handleChange({ target: { value: 'new value' } })

    expect(component.state.value).toBe('new value')
    expect(onChangeHandler).toHaveBeenCalledWith('text', 'new value')
  })
})
