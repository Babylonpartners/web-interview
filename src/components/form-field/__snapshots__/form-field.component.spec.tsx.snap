// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormField should match snapshot 1`] = `
<div
  className="form-field form-field--input"
>
  <div
    className="form-field__label"
  >
    <span
      className="icon icon--atom"
      title="atom"
    />
    <span>
      text
    </span>
  </div>
  <input
    className="form-field__input"
    onChange={[Function]}
  />
</div>
`;
